future.go

```diff
@@ -0,0 +1,158 @@
+package tarantool
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"gopkg.in/vmihailenco/msgpack.v2"
+)
+
+// Future is a handle for asynchronous request.
+type Future struct {
+	requestId   uint32
+	requestCode int32
+	timeout     time.Duration
+	resp        *Response
+	err         error
+	ready       chan struct{}
+	next        *Future
+	ctx         context.Context
+}
+
+// WithCtx sets a passed context to a Future object and
+// returns the same object with that context.
+func (fut *Future) WithCtx(ctx context.Context) *Future {
+	fut.ctx = ctx
+	return fut
+}
```

Comment:
Do you mean that we could keep the single context for all Future objects inside of a Connection object?

Reply:
In my opinion, it would look like a context for a Future object inside a Connection object, some kind of mapping. As a result , only the Connection will continue to manage lifetimes of futures.

But it may be hard to implement with the current implementation of request timeouts:

https://github.com/tarantool/go-tarantool/blob/1a1fe7b4e2ea77aa316ed4006aa445e512812c3c/connection.go#L892-L939

But there may be a way.

Reply:
Thank you for the clearing, sounds good. I come back soon with updates.

Reply:
I have tried a couple of approaches of cleaning. 
The first way was to run a separate goroutine which checks all the contexts of all Futures like it implemented with timeouts. I saw two options of implementing this:
- The first one was to go through all the futures and check their context (just like in `conn.timeouts()`). That approach has a disadvantage, that we have to use mutexes for locking Future queues (which are represented in `shard[n].requests`). And we can catch a situation, when some context of the future is already canceled or done, but we will handle it after we process N queues with other Futures. May be I've implement this bad, but I received a performance degradation by this one.
- The second one was to hold some slice of all futures with contexts and use the [reflect.SelectCase](https://pkg.go.dev/reflect#SelectCase) to listen if some context is done. Here I had some troubles of keeping the slice of `cases` in actual state, so that also damaged the performance with my implementation.

The second way is to let the Future object to have a pointer to it's Connection and ask for cleaning up, when the context is done. I understand that this approach makes the code architecture a bit dirtier, cause of sharing responsibility for the lifetime of a Future object between Connection and Future, but technically, only Connection object still has an ability to delete the Future. And that doesn't affect the performance at all.

If you have some ideas, please, share it. We could also discuss it by voice.

Reply:
You can try to process requests with contexts and old-style timeouts separately. Perhaps this will make things easier. And in general it will be better.

If you don't succeed it is possible to improve the current implementation. You can replace pointer to a connection by an interface. Something like this (be free to change naming and args):

```patch
+type futureCanceler interface {
+    Cancel(fut *Future, err error) error
+}
+
 // Future is a handle for asynchronous request.
 type Future struct {
     requestId uint32
@@ -19,7 +23,7 @@ type Future struct {
     ready     chan struct{}
     done      chan struct{}
     ctx       context.Context
-    conn      *Connection
+    canceler  futureCanceler
 }
```
You can add args to `NewFuture()` method and make it `NewFuture(context, canceler)`. After these changes, you will be able to write tests in `future_test.go` with mocking.

It will help us:
1. To define behavior of the Future in tests.
2. To break dependency Future <-> Connection. It will stay Future <- Connection.

Also please note the code. You should to test it too:
https://github.com/tarantool/go-tarantool/blob/1a1fe7b4e2ea77aa316ed4006aa445e512812c3c/connection.go#L701-L703